### DISCUSSION TITLE
When coupled with `git sparse-checkout`, should `git lfs prune` remove LFS files from deselected sparse filters?

### DISCUSSION THREAD

**Discussion by GarretFarrell** [git-lfs/git-lfs#5258] [Q&A] [answered]
Basically just the title.

`git sparse-checkout` seems like a massive win for the workflow that my team and I are designing, but we've hit a small snag.

The idea is that devs will move around the repo using `sparse-checkout` to work on large binary files in separate folders without having to keep a copy of the whole product at any one time. Product managers and CI will be able to checkout the full product for QA and build at the end.

Currently, each time you check out another folder to work on using `sparse-checkout`, LFS will grab the required assets as it should, but when you're finished, you deselect the `sparse-checkout` folder and run `git lfs prune` to clear out any unwanted data, but it doesn't clear out the objects relating to the folder you've just removed from your local repo.

This means that as you go through a project, your `.git/lfs/objects` directory will get bloated with all the files you've ever touched, even if you'll never need them again.

Is this a design decision, or is it just that `sparse-checkout` is new, and I should raise an issue for it in the future?

**Comment by bk2204** [comment, marked as answer]
Hey,

Git LFS doesn't by default prune objects which are required for recently checked out refs, regardless of sparsity patterns.  The philosophy is that if you've checked it out recently, you may want to check it out again soon.  You can however use `--recent` or `-f` to control prune data for other recent refs or even the checked out ref (respectively).  At least `-f` should work here, and if it doesn't, that's a bug.

There isn't an option to specifically respect sparsity patterns, although one could be added in the future if someone wants to contribute one.

**Reply by GarretFarrell** [reply]
Hi! Thanks for your quick response on this.

So, `--recent` doesn't take into account which sparse filters are active, so it has no effect in this case (perhaps because the checked-out ref is the same, regardless of sparsity pattern?) and `-f` just rips everything out, which isn't ideal given the intent, but it does work as advertised.

In your opinion, should the `--recent` flag take into account the sparsity filters as described? Or should it be a separate flag, called `--sparse` or something?

Also, in the meantime to get me over this hump, would it be catastrophic to get a script to run `git lfs ls-files --json`, run through each file, and any files that have all of their references `downloaded: true` and `checkout: false`, just delete them from `.git/lfs/objects`? Or is that going to cause untold grief down the line?

**Reply by bk2204** [reply]
Right, we don't consider sparse checkout at all.  Part of the problem is that, as you mentioned, you can change the sparsity patterns at will and we don't know what's been checked out as non-sparse in the past.  The goal of `--recent` is to keep things that were recently checked out, and in your case, some of the things you want to get rid of are indeed recent, even for your usage.  If they'd never been checked out, then they wouldn't have been downloaded.

If you wanted to add a flag `--sparse` that considers current sparsity patterns for `--recent` and ignores files that aren't in the sparsity pattern, then I don't think that would be a problem.

However, if you run the script you've proposed, you'll delete unpushed files, and as a result, you'll lose data.  You're much better off just using `-f` here, since that's much safer.

**Reply by GarretFarrell** [reply]
Understood! Thanks for the explanation.

The scary thing about `-f` is when I deselect _some_ of the sparsity patterns, but not all of them; if I use `-f` it gets rid of all LFS content as it should, but is that a risk for the LFS files I have left in my local when it comes to change tracking, discarding changes, committing, etc? Or is that known behaviour, and I'm just overthinking it?

I would instinctively want to replenish the objects in `.git/lfs/objects`, but `git lfs fetch` pulls the entire repo's LFS content, regardless of which sparsity patterns are active.

**Reply by bk2204** [reply]
`git lfs prune -f` doesn't remove files which are unpushed (unless you use the option which makes them not be verified on the server).  So if you've committed LFS files but haven't pushed them, then `-f` won't delete them.  However, it will delete files needed by the current checkout that are on the server, which will be downloaded again on demand.

We do want `git lfs fetch` and `git lfs pull` to honour sparse checkout patterns, but they don't yet.  If you wanted to implement that, that would be a very welcome change.

**Reply by GarretFarrell** [reply]
Thanks for your help on this; looks like `-f` will work for us until some kind of `git lfs prune --sparse` flag makes it a little more refined.

My Golang is rudimentary at best, so I'm probably not the best-qualified person, however, I can take a crack at it if I get time, and it can be refined after the fact, if that's helpful.

Would you like me to raise an issue for either "`git lfs pull`/`fetch` commands honour sparse checkout" or "add `--sparse` flag to `git lfs prune`"?

**Reply by bk2204** [reply]
I think the former is #3803.  An issue for the latter is of course welcome.

And if you end up wanting to do one or both of those things, we'll be happy to work with you to help refine a PR.  One of the nice things about Go is that it's relatively easy to pick up, so even if you're not super familiar with it, it shouldn't be too terribly difficult to figure out (and, as I said, we can help).

**Reply by GarretFarrell** [reply]
Great! I'll set up #5259 for `git lfs prune --sparse`, and I'll refer back to #3803 if I make any headway with it.

Thanks for all your help with this; I really appreciate it.
